Nearest neighbor particle interpolator

jperryhouts opened the pull request:
While you're all at it reviewing patches, here's one more!

This adds a particle interpolator which gets properties from the nearest particle in the same cell. If the cell is empty it gets the nearest particle from the nearest cell. This helps maintain sharp boundaries between compositional fields.

![cell_avg vs nearest_neighbor 3-2](https://cloud.githubusercontent.com/assets/6118266/25829945/6455bf2e-340e-11e7-9bcc-444242426b35.png)

tjhei commented:
/run-tests

bangerth commented:
This seems like something for @gassmoeller  :-)

gassmoeller commented:
I was on it yesterday, but got sidetracked again and again. Will try to finish the review now. @jperryhouts you will need to rebase to current master at some point, because the tester had a hickup yesterday.

gassmoeller commented on source/particle/interpolator/nearest_neighbor.cc:
```diff
@@ -0,0 +1,163 @@
+        const unsigned int n_particles = std::distance(particle_range.first,particle_range.second);
+        const unsigned int n_particle_properties = particles.begin()->second.get_properties().size();
+
+        std::vector<std::vector<double> > point_properties;
```
You already know how large this vector of vectors is going to be, and you can do all of the memory allocation here, instead of inside of the inner loops below the following way:
```
std::vector<double> temp(n_particle_properties,0.0);
point_properties (n_positions.size(),temp);
```
Then below you can always fill `point_properties[pos_idx][i]` instead of calling `push_back()`. This should speed things up significantly (I do not know how fast or slow this algorithm currently is, likely faster than cell average, but no reason to waste efficiency).

gassmoeller commented on source/particle/interpolator/nearest_neighbor.cc:
```diff
@@ -0,0 +1,163 @@
+                for (typename std::multimap<types::LevelInd, Particle<dim> >::const_iterator particle = particle_range.first;
+                     particle != particle_range.second; ++particle)
+                  {
+                    const double dist = positions[pos_idx].distance(particle->second.get_location());
```
This is way less important than my comment above.

gassmoeller commented on source/particle/interpolator/nearest_neighbor.cc:
```diff
@@ -0,0 +1,163 @@
+                        nearest_neighbor = particle->second;
+                      }
+                  }
+                std::vector<double> properties (n_particle_properties,0.0);
```
You will not need this vector anymore after the changes suggested above

gassmoeller commented on source/particle/interpolator/nearest_neighbor.cc:
```diff
@@ -0,0 +1,163 @@
+                        nearest_neighbor_cell = i;
+                      }
+                  }
+                std::vector<double> properties (n_particle_properties,0.0);
```
same here. no need after the changes

gassmoeller commented on source/particle/interpolator/nearest_neighbor.cc:
```diff
@@ -0,0 +1,163 @@
+                      }
+                  }
+                std::vector<double> properties (n_particle_properties,0.0);
+                const std::vector<double> neighbor_props = properties_at_points(particles,
```
I think you can just assign the return value to `point_properties`. It is the job of the recursively called function to select the appropriate components already.

jperryhouts commented:
Thanks @gassmoeller for the quick (and specific) review! I addressed all the comments, so I think this should be ready to go as long as the tester is happy with it.

gassmoeller commented:
I just remembered, could you write an entry for modules/changes? This is a contribution that definitely deserves an entry over there.
